folding-schemes/src/folding/ova/nifs.rs

```diff
@@ -0,0 +1,498 @@
+/// This module contains the implementation of the Ova scheme NIFS (Non-Interactive Folding Scheme) as
+/// outlined in the protocol description doc: <https://hackmd.io/V4838nnlRKal9ZiTHiGYzw?both#Construction>
+/// authored by Benedikt Bünz.
+use ark_crypto_primitives::sponge::Absorb;
+use ark_ec::{CurveGroup, Group};
+use ark_std::One;
+use std::marker::PhantomData;
+
+use super::{CommittedInstance, Witness};
+use crate::arith::r1cs::R1CS;
+use crate::commitment::CommitmentScheme;
+use crate::transcript::Transcript;
+use crate::utils::vec::{hadamard, mat_vec_mul, vec_scalar_mul, vec_sub};
+use crate::Error;
+
+/// Implements all the operations executed by the Non-Interactive Folding Scheme described in the protocol
+/// spec by Bünz in the [original HackMD](https://hackmd.io/V4838nnlRKal9ZiTHiGYzw?both#Construction).
+/// `H` specifies whether the NIFS will use a blinding factor
+pub struct NIFS<C: CurveGroup, CS: CommitmentScheme<C, H>, const H: bool = false> {
+    _c: PhantomData<C>,
+    _cp: PhantomData<CS>,
+}
+
+impl<C: CurveGroup, CS: CommitmentScheme<C, H>, const H: bool> NIFS<C, CS, H>
+where
+    <C as Group>::ScalarField: Absorb,
+{
+    /// Computes the T parameter (Cross Terms) as in Nova.
+    /// The wrapper is only in place to facilitate the calling as we need
+    /// to reconstruct the `z`s being folded in order to compute T.
+    pub fn compute_T(
+        r1cs: &R1CS<C::ScalarField>,
+        w_i: &Witness<C>,
+        x_i: &[C::ScalarField],
+        W_i: &Witness<C>,
+        X_i: &[C::ScalarField],
+        mu: C::ScalarField,
+    ) -> Result<Vec<C::ScalarField>, Error> {
+        crate::folding::nova::nifs::NIFS::<C, CS, H>::compute_T(
+            r1cs,
+            C::ScalarField::one(),
+            mu,
+            &[vec![C::ScalarField::one()], x_i.to_vec(), w_i.w.to_vec()].concat(),
+            &[vec![mu], X_i.to_vec(), W_i.w.to_vec()].concat(),
+        )
+    }
+
+    /// Computes the E parameter (Error Terms) as in Nova.
+    /// The wrapper is only in place to facilitate the calling as we need
+    /// to reconstruct the `z`s being folded in order to compute E.
+    ///
+    /// Not only that, but notice that the incoming-instance `mu` parameter is always
+    /// equal to 1. Therefore, we can save the some computations.
+    pub fn compute_E(
+        r1cs: &R1CS<C::ScalarField>,
+        W_i: &Witness<C>,
+        X_i: &[C::ScalarField],
+        mu: C::ScalarField,
+    ) -> Result<Vec<C::ScalarField>, Error> {
+        let (A, B, C) = (r1cs.A.clone(), r1cs.B.clone(), r1cs.C.clone());
+
+        let z_prime = [&[mu], X_i, &W_i.w].concat();
+        // this is parallelizable (for the future)
+        let Az_prime = mat_vec_mul(&A, &z_prime)?;
+        let Bz_prime = mat_vec_mul(&B, &z_prime)?;
+        let Cz_prime = mat_vec_mul(&C, &z_prime)?;
+
+        let Az_prime_Bz_prime = hadamard(&Az_prime, &Bz_prime)?;
+        let muCz_prime = vec_scalar_mul(&Cz_prime, &mu);
```

Comment:
here it says that "mu is always 1 therefore we can save some computations", but in fact mu might not be 1, and also the code of the method is using mu.
But as with https://github.com/privacy-scaling-explorations/sonobe/pull/163#discussion_r1785152424, don't mind if it stays as it is, since it is fixed already in #165